Amazon sales data
|Column name | Description |
|---|---|
|Order ID| Unique ID for every order. (ID)|
|Date| Date of the sale. (Date)|
|Status| Status of the sale. (String)|
|Fulfilment| Method of fulfilment. (String)|
|Sales Channel| Channel used on sale. (String)|
|ship-service-level| Shipping Category. (String)|
|Style| Style of the product. (String)|
|SKU| Stock Keeping Unit. (String)|
|Category| Type of product. (String)|
|Size|Size of the product. (String)|
|ASIN| Amazon Standard Identification Number. (String)|
|Courier Status| Status of the courier. (String)|
|Qty| Quantity of the product. (Integer)|
|currency| The currency used for the sale. (String)|
|Amount| Amount of the sale. (Float)|
|ship-city| City of ship address. (String)|
|ship-state| State of ship address. (String)|
|ship-postal-code| Postcode of ship address. (String)|
|ship-country| Country of ship address. (String)|
|promotion-ids| Applied promo on sale. (String)|
|B2B| Business to business sale. (Boolean)|
|fulfilled-by| Amazon easy ship service. (String)|
|Unnamed: 22| unknown|

Project details \
Python version : 3.11.6 \
Dataset source : https://www.kaggle.com/datasets/thedevastator/unlock-profits-with-e-commerce-sales-data



## Import Dataset

Here we using pandas package for read csv datafile.
Name of the dataset is "Amazon Sale Report.csv"

In [1]:
#import pandas package
import pandas as pd

#because we have 23 columns, we need to set max columns option to None (max) so that the table is not cut off
pd.set_option('display.max_columns', None)

#read csv file
sales_data = pd.read_csv('Amazon Sale Report.csv',low_memory=False,index_col='index')

#check result
sales_data.head()

Order ID      Date                        Status Fulfilment  \
index                                                                           
0      405-8078784-5731545  04-30-22                     Cancelled   Merchant   
1      171-9198151-1101146  04-30-22  Shipped - Delivered to Buyer   Merchant   
2      404-0687676-7273146  04-30-22                       Shipped     Amazon   
3      403-9615377-8133951  04-30-22                     Cancelled   Merchant   
4      407-1069790-7240320  04-30-22                       Shipped     Amazon   

      Sales Channel  ship-service-level    Style              SKU  \
index                                                               
0          Amazon.in           Standard   SET389   SET389-KR-NP-S   
1          Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL   
2          Amazon.in          Expedited  JNE3371    JNE3371-KR-XL   
3          Amazon.in           Standard    J0341       J0341-DR-L   
4          Amazon.in          Expedited  JNE3671  JNE3671-TU-XXXL   

            Category Size        ASIN Courier Status  Qty currency  Amount  \
index                                                                        
0                Set    S  B09KXVBD7Z            NaN    0      INR  647.62   
1              kurta  3XL  B09K3WFS32        Shipped    1      INR  406.00   
2              kurta   XL  B07WV4JV4D        Shipped    1      INR  329.00   
3      Western Dress    L  B099NRCT7B            NaN    0      INR  753.33   
4                Top  3XL  B098714BZP        Shipped    1      INR  574.00   

         ship-city   ship-state  ship-postal-code ship-country  \
index                                                            
0           MUMBAI  MAHARASHTRA          400081.0           IN   
1        BENGALURU    KARNATAKA          560085.0           IN   
2      NAVI MUMBAI  MAHARASHTRA          410210.0           IN   
3       PUDUCHERRY   PUDUCHERRY          605008.0           IN   
4          CHENNAI   TAMIL NADU          600073.0           IN   

                                           promotion-ids    B2B fulfilled-by  \
index                                                                          
0                                                    NaN  False    Easy Ship   
1      Amazon PLCC Free-Financing Universal Merchant ...  False    Easy Ship   
2           IN Core Free Shipping 2015/04/08 23-48-5-108   True          NaN   
3                                                    NaN  False    Easy Ship   
4                                                    NaN  False          NaN   

      Unnamed: 22  
index              
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

In [2]:
#check null and dtype from every feature
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 128975 entries, 0 to 128974
Data columns (total 23 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Order ID            128975 non-null  object 
 1   Date                128975 non-null  object 
 2   Status              128975 non-null  object 
 3   Fulfilment          128975 non-null  object 
 4   Sales Channel       128975 non-null  object 
 5   ship-service-level  128975 non-null  object 
 6   Style               128975 non-null  object 
 7   SKU                 128975 non-null  object 
 8   Category            128975 non-null  object 
 9   Size                128975 non-null  object 
 10  ASIN                128975 non-null  object 
 11  Courier Status      122103 non-null  object 
 12  Qty                 128975 non-null  int64  
 13  currency            121180 non-null  object 
 14  Amount              121180 non-null  float64
 15  ship-city           128942 non-null  ob

from above we know some feature is not in correct dtype. Let's fix date column first.

change Date column from object --> datetime \
with format MM-DD-YY

In [3]:
#change dtype of column 'Date'
sales_data['Date']=pd.to_datetime(sales_data['Date'],format='%m-%d-%y')

#check result
sales_data['Date'].info()

<class 'pandas.core.series.Series'>
Index: 128975 entries, 0 to 128974
Series name: Date
Non-Null Count   Dtype         
--------------   -----         
128975 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 2.0 MB


Next because we want to calculate amount of sale. First we need to check the currency taht we have on or data.

In [4]:
#Check all currency type that we have
#show as count
display(sales_data['currency'].value_counts(dropna=False))

#show on percent
sales_data['currency'].value_counts(dropna=False,normalize=True)*100

currency
INR    121180
NaN      7795
Name: count, dtype: int64

currency
INR    93.956193
NaN     6.043807
Name: proportion, dtype: float64

Here we got 2 unique value from our data, INR and NaN, Where NaN is null value.

Let's check whether there is a significant difference in value between data with currency INR and NaN.

In [5]:
#slice data with currency INR for Amount column only
cek_INR = sales_data[['currency','Amount']][sales_data['currency']=='INR']

#slice data with currency NaN (not INR) for Amount column only
cek_NaN = sales_data[['currency','Amount']][sales_data['currency']!='INR']

#check result with general stat
display(cek_INR.describe())
cek_NaN.describe()

Amount
count  121180.000000
mean      648.561465
std       281.211687
min         0.000000
25%       449.000000
50%       605.000000
75%       788.000000
max      5584.000000

Amount
count     0.0
mean      NaN
std       NaN
min       NaN
25%       NaN
50%       NaN
75%       NaN
max       NaN

It turns out, data with Nan's currency does not have a nominal value in the Amount column. Why is that? let's confirm further.

In [6]:
#slice data with currency NaN (not INR) on all column we have
cek_NaN_con = sales_data[sales_data['currency']!='INR']

#cek result
cek_NaN_con.head(10)

Order ID       Date     Status Fulfilment Sales Channel   \
index                                                                        
8      407-5443024-5233168 2022-04-30  Cancelled     Amazon      Amazon.in   
29     404-5933402-8801952 2022-04-30  Cancelled   Merchant      Amazon.in   
65     171-4137548-0481151 2022-04-30  Cancelled     Amazon      Amazon.in   
84     403-9950518-0349133 2022-04-30  Cancelled     Amazon      Amazon.in   
95     405-9112089-3379536 2022-04-30  Cancelled     Amazon      Amazon.in   
101    402-7841951-1872363 2022-04-30  Cancelled     Amazon      Amazon.in   
132    171-0293372-4913973 2022-04-30  Cancelled     Amazon      Amazon.in   
139    403-6442534-2769112 2022-04-30  Cancelled     Amazon      Amazon.in   
155    402-8078642-8565124 2022-04-30  Cancelled     Amazon      Amazon.in   
165    406-0218328-1773963 2022-04-30  Cancelled     Amazon      Amazon.in   

      ship-service-level     Style                  SKU Category Size  \
index                                                                   
8              Expedited    SET200  SET200-KR-NP-A-XXXL      Set  3XL   
29              Standard   JNE2132  JNE2132-KR-398-XXXL    kurta  3XL   
65             Expedited   JNE3373       JNE3373-KR-XXL    kurta  XXL   
84             Expedited   JNE3510         JNE3510-KR-M    kurta    M   
95             Expedited   JNE3405         JNE3405-KR-L    kurta    L   
101            Expedited    SET347       SET347-KR-NP-S      Set    S   
132            Expedited    SET269    SET269-KR-NP-XXXL      Set  3XL   
139            Expedited     J0300        J0300-TP-XXXL      Top  3XL   
155            Expedited   JNE3887        JNE3887-KR-XL    kurta   XL   
165            Expedited  PJNE2100    PJNE2100-KR-N-6XL    kurta  6XL   

             ASIN Courier Status  Qty currency  Amount  ship-city  \
index                                                               
8      B08L91ZZXN      Cancelled    0      NaN     NaN  HYDERABAD   
29     B07JG3CND8            NaN    0      NaN     NaN   GUWAHATI   
65     B082W8RWN1      Cancelled    0      NaN     NaN      Dahod   
84     B08WPR5MCB      Cancelled    0      NaN     NaN  HYDERABAD   
95     B081WSCKPQ      Cancelled    0      NaN     NaN       PUNE   
101    B09RKDJ9SQ      Cancelled    0      NaN     NaN  NEW DELHI   
132    B0983FSRFS      Cancelled    0      NaN     NaN      KOLAR   
139    B099S7RDLZ      Cancelled    0      NaN     NaN   AZAMGARH   
155    B09TZVXQ1J      Cancelled    0      NaN     NaN     Puri-2   
165    B09LD3PMBT      Cancelled    0      NaN     NaN  WALAJAPET   

          ship-state  ship-postal-code ship-country  \
index                                                 
8          TELANGANA          500008.0           IN   
29             ASSAM          781003.0           IN   
65           Gujarat          389151.0           IN   
84         TELANGANA          500072.0           IN   
95       MAHARASHTRA          411046.0           IN   
101            DELHI          110036.0           IN   
132        KARNATAKA          563101.0           IN   
139    UTTAR PRADESH          276302.0           IN   
155           ODISHA          752002.0           IN   
165       TAMIL NADU          632513.0           IN   

                                      promotion-ids    B2B fulfilled-by  \
index                                                                     
8      IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   
29                                              NaN  False    Easy Ship   
65                                              NaN  False          NaN   
84                                              NaN  False          NaN   
95                                              NaN  False          NaN   
101                                             NaN  False          NaN   
132                                             NaN  False          NaN   
139                                    

Of the 10 existing data, at first glance what appears to be sales transaction data has the status 'Cancelled'. But, let's make sure again.

In [7]:
#count 
cek_NaN_con['Status'].value_counts(dropna=False)

Status
Cancelled                       7566
Shipped                          208
Shipped - Delivered to Buyer       8
Shipping                           8
Shipped - Returned to Seller       3
Pending                            2
Name: count, dtype: int64

It turns out that it's not just transactions that are cancelled, there is also transaction data with other statuses.

In [8]:
#we want to know every row data with different Status, we will take 2 sample
cek_NaN_con.groupby('Status', group_keys=False).apply(lambda df: df.sample(2))

Order ID       Date                        Status  \
index                                                                  
44486   404-4392343-0193134 2022-04-03                     Cancelled   
69863   171-6623081-8091558 2022-05-15                     Cancelled   
91209   407-4176879-6571509 2022-06-29                       Pending   
58986   403-7066611-8150715 2022-05-24                       Pending   
118088  S02-0219787-4779903 2022-06-08                       Shipped   
100239  407-9261873-2421156 2022-06-21                       Shipped   
35337   407-9335911-1673924 2022-04-09  Shipped - Delivered to Buyer   
63344   404-3955601-0129966 2022-05-20  Shipped - Delivered to Buyer   
74296   171-0040661-6643541 2022-05-11  Shipped - Returned to Seller   
84211   171-3459992-9664354 2022-05-04  Shipped - Returned to Seller   
107949  S02-9599483-2736812 2022-06-15                      Shipping   
107951  S02-9218933-0457140 2022-06-15                      Shipping   

       Fulfilment Sales Channel  ship-service-level    Style  \
index                                                          
44486      Amazon      Amazon.in          Expedited  JNE3833   
69863    Merchant      Amazon.in           Standard  JNE3501   
91209      Amazon      Amazon.in          Expedited  JNE3486   
58986      Amazon      Amazon.in          Expedited   SET145   
118088     Amazon     Non-Amazon           Standard  JNE3364   
100239     Amazon      Amazon.in          Expedited   SET308   
35337    Merchant      Amazon.in           Standard  MEN5028   
63344    Merchant      Amazon.in           Standard    J0151   
74296    Merchant      Amazon.in           Standard   SET401   
84211    Merchant      Amazon.in           Standard  JNE3679   
107949     Amazon     Non-Amazon           Standard    J0012   
107951     Amazon     Non-Amazon           Standard    J0009   

                        SKU Category Size        ASIN Courier Status  Qty  \
index                                                                       
44486        JNE3833-KR-XXL    kurta  XXL  B09PD6BR4C      Cancelled    0   
69863          JNE3501-KR-S    kurta    S  B08MYS1L63            NaN    0   
91209        JNE3486-KR-XXL    kurta  XXL  B08W8K156H      Cancelled    0   
58986        SET145-KR-NP-L      Set    L  B0822SX9WB      Cancelled    0   
118088  JNE3364-KR-1051-A-L    kurta    L  B07WP5G7MK      Unshipped    1   
100239       SET308-KR-PP-S      Set    S  B09K3MP21Q      Cancelled    0   
35337         MEN5028-KR-XL    kurta   XL  B08YYTL6BX            NaN    0   
63344         J0151-KR-A-XL    kurta   XL  B08N19YB92            NaN    0   
74296        SET401-KR-NP-L      Set    L  B09VC5ZGD8            NaN    0   
84211          JNE3679-TU-L      Top    L  B0943J4SL3            NaN    0   
107949          J0012-SKD-L      Set    L  B0894XGSH9      Unshipped    1   
107951          J0009-SKD-M      Set    M  B0894Y7ZF8      Unshipped    1   

       currency  Amount    ship-city      ship-state  ship-postal-code  \
index                                                                    
44486       NaN     NaN       MUMBAI     MAHARASHTRA          400077.0   
69863       NaN     NaN      LUCKNOW   UTTAR PRADESH          226024.0   
91209       NaN     NaN      NANDYAL  ANDHRA PRADESH          518502.0   
58986       NaN     NaN         PUNE     MAHARASHTRA          411021.0   
118088      NaN     NaN        Surat         Gujarat          394210.0   
100239      NaN     NaN       JAIPUR       RAJASTHAN          302015.0   
35337       NaN     NaN  GANDHINAGAR         Gujarat          382007.0   
63344       NaN     NaN      PATIALA          PUNJAB          147001.0   
74296       NaN     NaN        ELURU  ANDHRA PRADESH          534001.0   
84211       NaN     NaN      HOSKOTE       KARNATAKA          562114.0   
107949      NaN     NaN        Surat         Gujarat          394210.0   
107951      NaN     NaN        Surat         Gujarat          394210.0  

From the data displayed above, it turns out that there is another column that determines the status of the sales transaction, namely "Courier Status".

For example, in the Status column it is stated as 'Shipped' but in the Courier Status column it is stated as 'Cancelled' and there is also something stated as 'Unshipped', which means these two columns could be in conflict.

lets check unique combination for every values on "Status" and "Courier Status" columns.

In [9]:
cek_NaN_con[['Status','Courier Status']].value_counts().sort_index()

Status     Courier Status
Cancelled  Cancelled         5840
           Unshipped            1
Pending    Cancelled            2
Shipped    Cancelled           93
           Unshipped          115
Shipping   Unshipped            8
Name: count, dtype: int64

From above, now we got the schema for our data:

|Status|Courier Status| riddle|
|---|---|---|
|Cancelled|Cancelled| Normal condition, transaction is canceled so Amount and currency is null|
|Cancelled|Unshipped| Maybe the product has been packaged and ready to be sent, but it turns out the transaction has been canceled and the goods are already on their way to the expedition's cargo warehouse but have not yet been sent to their destination.
etc.

With conditions like this, we can conclude that the data rows whose 'currency' column is NaN are data rows from transactions that were canceled under any cancellation conditions.

because currently what we need is to calculate the value from the "Amount" column to get the daily transaction value, then based on the results of previous exploration, we need to remove this data from the dataset that we will use in creating the forecasting model.

Maybe we can add a new column "Success" that can be used as an indicator of whether the transaction was successful or not.

In [10]:
#apply mapping by condition
sales_data['Success'] = sales_data['Amount'].apply(lambda x: 'No' if pd.isnull(x) else 'Yes')

#check result
sales_data.groupby('Success', group_keys=False).apply(lambda df: df.sample(3))

Order ID       Date     Status Fulfilment Sales Channel   \
index                                                                         
84568   404-6996095-6009132 2022-05-04  Cancelled     Amazon      Amazon.in   
120273  403-3805456-9818708 2022-06-06  Cancelled     Amazon      Amazon.in   
85928   408-4412777-5685133 2022-05-03  Cancelled     Amazon      Amazon.in   
25089   403-6911566-0461118 2022-04-15    Shipped     Amazon      Amazon.in   
128068  408-3553357-8133961 2022-06-01    Shipped     Amazon      Amazon.in   
97584   171-6506110-4873904 2022-06-23    Shipped     Amazon      Amazon.in   

       ship-service-level    Style              SKU Category Size        ASIN  \
index                                                                           
84568           Expedited  JNE3805     JNE3805-KR-L    kurta    L  B09K3ZKHMW   
120273          Expedited    J0351    J0351-SET-XXL      Set  XXL  B09CT76R4C   
85928           Expedited  MEN5032     MEN5032-KR-S    kurta    S  B08YYXJJ4V   
25089           Expedited   SET073  SET073-KR-SHA-S      Set    S  B07V5WFLND   
128068          Expedited  JNE3837     JNE3837-KR-M    kurta    M  B09RKCMWVM   
97584           Expedited   SET400   SET400-KR-NP-L      Set    L  B09W644VYK   

       Courier Status  Qty currency  Amount              ship-city  \
index                                                                
84568       Cancelled    0      NaN     NaN              KALABURGI   
120273      Cancelled    0      NaN     NaN             CHANDIGARH   
85928       Cancelled    0      NaN     NaN  MANDVI SURAT DISTRICT   
25089         Shipped    1      INR   824.0             JAMSHEDPUR   
128068        Shipped    1      INR   533.0                  ALWAR   
97584         Shipped    1      INR  1068.0      GOBICHETTIPALAYAM   

        ship-state  ship-postal-code ship-country  \
index                                               
84568    KARNATAKA          585102.0           IN   
120273  CHANDIGARH          160012.0           IN   
85928      Gujarat          394160.0           IN   
25089    JHARKHAND          831012.0           IN   
128068   RAJASTHAN          301001.0           IN   
97584   TAMIL NADU          638452.0           IN   

                                       promotion-ids    B2B fulfilled-by  \
index                                                                      
84568                                            NaN  False          NaN   
120273  IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   
85928                                            NaN  False          NaN   
25089   IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   
128068                                           NaN  False          NaN   
97584   IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   

       Unnamed: 22 Success  
index                       
84568        False      No  
120273       False      No  
85928        False      No  
25089          NaN     Yes  
128068       False     Yes  
97584        False     Yes

Let's check another null value on our data.

In [11]:
#count null on all columns
sales_data.isna().sum()

Order ID                  0
Date                      0
Status                    0
Fulfilment                0
Sales Channel             0
ship-service-level        0
Style                     0
SKU                       0
Category                  0
Size                      0
ASIN                      0
Courier Status         6872
Qty                       0
currency               7795
Amount                 7795
ship-city                33
ship-state               33
ship-postal-code         33
ship-country             33
promotion-ids         49153
B2B                       0
fulfilled-by          89698
Unnamed: 22           49050
Success                   0
dtype: int64

Another null value found on 8 columns. But it seems that only the fields related to the shipping address we need to pay attention to, the other fields look normal if they are empty.

We will check by ship-country column.

In [12]:
sales_data[~pd.notna(sales_data['ship-country'])].head()

Order ID       Date     Status Fulfilment Sales Channel   \
index                                                                        
1871   404-0566904-4825137 2022-04-29    Shipped     Amazon      Amazon.in   
1872   404-0566904-4825137 2022-04-29    Shipped     Amazon      Amazon.in   
8752   406-4003386-8768363 2022-04-25    Shipped     Amazon      Amazon.in   
11215  402-0107720-7057168 2022-04-23    Shipped     Amazon      Amazon.in   
13252  407-4532637-8415521 2022-04-22  Cancelled   Merchant      Amazon.in   

      ship-service-level    Style           SKU Category Size        ASIN  \
index                                                                       
1871           Expedited    J0301    J0301-TP-L      Top    L  B099S6795L   
1872           Expedited  JNE3697  JNE3697-KR-L    kurta    L  B09811Y7WM   
8752           Expedited  JNE3376  JNE3376-KR-M    kurta    M  B082W7HW86   
11215          Expedited    J0003   J0003-SET-S      Set    S  B0894X27FC   
13252           Standard  JNE3405  JNE3405-KR-S    kurta    S  B081WX4G4Q   

      Courier Status  Qty currency  Amount ship-city ship-state  \
index                                                             
1871         Shipped    1      INR   493.0       NaN        NaN   
1872         Shipped    1      INR   458.0       NaN        NaN   
8752         Shipped    1      INR   432.0       NaN        NaN   
11215        Shipped    1      INR   654.0       NaN        NaN   
13252            NaN    0      INR   380.0       NaN        NaN   

       ship-postal-code ship-country  \
index                                  
1871                NaN          NaN   
1872                NaN          NaN   
8752                NaN          NaN   
11215               NaN          NaN   
13252               NaN          NaN   

                                      promotion-ids    B2B fulfilled-by  \
index                                                                     
1871   IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   
1872   IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   
8752                                            NaN  False          NaN   
11215  IN Core Free Shipping 2015/04/08 23-48-5-108  False          NaN   
13252                                           NaN  False    Easy Ship   

      Unnamed: 22 Success  
index                      
1871          NaN     Yes  
1872          NaN     Yes  
8752          NaN     Yes  
11215         NaN     Yes  
13252         NaN     Yes

It was shipped but the shipping addres is null. Maybe its was because system error.

Because the number is very small and we already have quite a lot of success data with 'Yes' value, then we can just delete it from our data

In [13]:
#drop null value on ship-country column
sales_data_fix = sales_data.dropna(subset='ship-country')

#count null value on data
sales_data_fix.isna().sum()

Order ID                  0
Date                      0
Status                    0
Fulfilment                0
Sales Channel             0
ship-service-level        0
Style                     0
SKU                       0
Category                  0
Size                      0
ASIN                      0
Courier Status         6869
Qty                       0
currency               7793
Amount                 7793
ship-city                 0
ship-state                0
ship-postal-code          0
ship-country              0
promotion-ids         49145
B2B                       0
fulfilled-by          89678
Unnamed: 22           49037
Success                   0
dtype: int64

OK, our data looks fine right now.

## Find Problem

One of the things that could be a detrimental opportunity for an e-commerce business is the failure of a transaction, especially in the buying and selling process, the product must be sent to the customer without the need for direct buying and selling interaction. This provides the opportunity for fictitious transactions or orders to be canceled while they are in the delivery process.

To avoid transactions being cancelled when they have entered the delivery process, this can be overcome by making regulations not to allow cancellation of orders that have been sent, but this regulation should have been implemented. However, there may be certain cases where it is still permissible to do so. Back to the cost of doing business.

But can we predict these failed transactions earlier? \
so that we can implement precautionary measures such as eliminating cash on delivery payment methods (if any), or coming up with captchas to verify whether transactions are made by humans, or so on.

## Machine Learning 1

Our first model is a model to predict whether a transaction will succeed or not.

Since machines cannot not understand words, we need to perform feature engineering on our data that is still in the form of text and objects first.

Next, we need to remove other column that we don't need it, from the description in we know we can't use columns:
- Order ID
- Unnamed: 22
- Date
- SKU
- ASIN

than we know all transaction the currency is in INR, so we dont need that.
- currency
- Status
- Courier Status

lets check another column like : Sales Channel and ship-country

In [14]:
display(sales_data_fix['Sales Channel '].value_counts(dropna=False))
sales_data_fix['ship-country'].value_counts(dropna=False)

Sales Channel 
Amazon.in     128818
Non-Amazon       124
Name: count, dtype: int64

ship-country
IN    128942
Name: count, dtype: int64

It turns out that the ship-country column has the same value in every row of data we have. Then this column that we need to add to be removed. Than, because we have postal code, we dont need other ship address like ship-city, ship-state.

Next we check "Category" and "Style" column, does it have many combination?

In [15]:
#Check combination of Category to style
cek_NaN_con[['Category','Style']].value_counts().sort_index()

Category  Style   
Blouse    BL003       1
          BL006       1
          BL013       2
          BL017       1
          BL020       1
                     ..
kurta     PJNE3421    2
          PJNE3423    1
          PJNE3440    1
          PJNE3445    1
          PJNE3568    4
Name: count, Length: 938, dtype: int64

we got 938 combination of category by style. We dont need that, we will only using Category column.

In [16]:
#drop columns that not needed
data_1 = sales_data_fix.drop(['Order ID','Unnamed: 22','Date','SKU','ASIN','currency','Status','Courier Status', 'ship-country','ship-city', 'ship-state','Style'],axis=1)

#reset index
data_1.reset_index(inplace=True,drop=True)

data_1['ship-postal-code'] = data_1['ship-postal-code'].astype('int')

#check result
data_1.head()

Fulfilment Sales Channel  ship-service-level       Category Size  Qty  \
0   Merchant      Amazon.in           Standard            Set    S    0   
1   Merchant      Amazon.in           Standard          kurta  3XL    1   
2     Amazon      Amazon.in          Expedited          kurta   XL    1   
3   Merchant      Amazon.in           Standard  Western Dress    L    0   
4     Amazon      Amazon.in          Expedited            Top  3XL    1   

   Amount  ship-postal-code  \
0  647.62            400081   
1  406.00            560085   
2  329.00            410210   
3  753.33            605008   
4  574.00            600073   

                                       promotion-ids    B2B fulfilled-by  \
0                                                NaN  False    Easy Ship   
1  Amazon PLCC Free-Financing Universal Merchant ...  False    Easy Ship   
2       IN Core Free Shipping 2015/04/08 23-48-5-108   True          NaN   
3                                                NaN  False    Easy Ship   
4                                                NaN  False          NaN   

  Success  
0     Yes  
1     Yes  
2     Yes  
3     Yes  
4     Yes

Next, because we have some categorical column with unordered data in our dataset. We need to transform it to onehot encoder or split it to new column. Here we use pandas.get_dummies to do that thing.

In [17]:
#categorical column list
columns = ['Fulfilment','Sales Channel ','Category']

#using loop to get dummies than drop original column and join the dummies to dataset with column name prefix is the original column name
for i in columns:
  dum_df = pd.get_dummies(data_1[i],dtype='int',prefix=i)
  data_1 = data_1.drop(columns=i).join(dum_df)

data_1['Sales Channel Amazon'] = data_1['Sales Channel _Amazon.in'].copy()
data_1['Sales Channel Non-Amazon'] = data_1['Sales Channel _Non-Amazon'].copy()
data_1.drop(['Sales Channel _Amazon.in','Sales Channel _Non-Amazon'],axis=1,inplace=True)

data_1.head()

ship-service-level Size  Qty  Amount  ship-postal-code  \
0           Standard    S    0  647.62            400081   
1           Standard  3XL    1  406.00            560085   
2          Expedited   XL    1  329.00            410210   
3           Standard    L    0  753.33            605008   
4          Expedited  3XL    1  574.00            600073   

                                       promotion-ids    B2B fulfilled-by  \
0                                                NaN  False    Easy Ship   
1  Amazon PLCC Free-Financing Universal Merchant ...  False    Easy Ship   
2       IN Core Free Shipping 2015/04/08 23-48-5-108   True          NaN   
3                                                NaN  False    Easy Ship   
4                                                NaN  False          NaN   

  Success  Fulfilment_Amazon  Fulfilment_Merchant  Category_Blouse  \
0     Yes                  0                    1                0   
1     Yes                  0                    1                0   
2     Yes                  1                    0                0   
3     Yes                  0                    1                0   
4     Yes                  1                    0                0   

   Category_Bottom  Category_Dupatta  Category_Ethnic Dress  Category_Saree  \
0                0                 0                      0               0   
1                0                 0                      0               0   
2                0                 0                      0               0   
3                0                 0                      0               0   
4                0                 0                      0               0   

   Category_Set  Category_Top  Category_Western Dress  Category_kurta  \
0             1             0                       0               0   
1             0             0                       0               1   
2             0             0                       0               1   
3             0             0                       1               0   
4             0             1                       0               0   

   Sales Channel Amazon  Sales Channel Non-Amazon  
0                     1                         0  
1                     1                         0  
2                     1                         0  
3                     1                         0  
4                     1                         0

Than, we need to apply some mapping on another columns

In [18]:
#check unique value and count it
display(data_1['ship-service-level'].value_counts(dropna=False))
display(data_1['Size'].value_counts(dropna=False))
display(data_1['promotion-ids'].value_counts(dropna=False))

ship-service-level
Expedited    88595
Standard     40347
Name: count, dtype: int64

Size
M       22704
L       22123
XL      20872
XXL     18093
S       17084
3XL     14815
XS      11160
6XL       738
5XL       550
4XL       425
Free      378
Name: count, dtype: int64

promotion-ids
NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       

we can use feature engineering for the ship-service-level and Size columns because they have consecutive values. As for the promotion-ids column, it seems like it should be replaced by stating it as a column whether this transaction uses a promo or not, so we will change it to boolean.

In [19]:
#apply function to label encoder ship-service-level
data_1['ship-service-level'] = data_1['ship-service-level'].apply(lambda x : 1 if x=='Standard' else 2)

#apply map to Size column by size dictionary
data_1['Size'] = data_1['Size'].map({'Free':0,'XS':1,'S':2,'M':3,'L':4,'XL':5,'XXL':6,'3XL':7,'4XL':8,'5XL':9,'6XL':10})

#apply if function to promotion-id, change it to boolean data and store it in promotion column
data_1['promotion'] = data_1['promotion-ids'].apply(lambda x: 0 if pd.isnull(x) else 1)

#drop promotion-id column
data_1.drop('promotion-ids',axis=1, inplace=True)

#fix other columns
data_1['B2B'] = data_1['B2B'].astype('int')
data_1['Success'] = data_1['Success'].apply(lambda x : 1 if x=='Yes' else 0)
data_1['fulfilled-by'] = data_1['fulfilled-by'].apply(lambda x : 1 if x=='Easy Ship' else 0)

#check result
display(data_1.info())
data_1.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128942 entries, 0 to 128941
Data columns (total 22 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ship-service-level        128942 non-null  int64  
 1   Size                      128942 non-null  int64  
 2   Qty                       128942 non-null  int64  
 3   Amount                    121149 non-null  float64
 4   ship-postal-code          128942 non-null  int64  
 5   B2B                       128942 non-null  int64  
 6   fulfilled-by              128942 non-null  int64  
 7   Success                   128942 non-null  int64  
 8   Fulfilment_Amazon         128942 non-null  int64  
 9   Fulfilment_Merchant       128942 non-null  int64  
 10  Category_Blouse           128942 non-null  int64  
 11  Category_Bottom           128942 non-null  int64  
 12  Category_Dupatta          128942 non-null  int64  
 13  Category_Ethnic Dress     128942 non-null  i

None

ship-service-level  Size  Qty  Amount  ship-postal-code  B2B  fulfilled-by  \
0                   1     2    0  647.62            400081    0             1   
1                   1     7    1  406.00            560085    0             1   
2                   2     5    1  329.00            410210    1             0   
3                   1     4    0  753.33            605008    0             1   
4                   2     7    1  574.00            600073    0             0   

   Success  Fulfilment_Amazon  Fulfilment_Merchant  Category_Blouse  \
0        1                  0                    1                0   
1        1                  0                    1                0   
2        1                  1                    0                0   
3        1                  0                    1                0   
4        1                  1                    0                0   

   Category_Bottom  Category_Dupatta  Category_Ethnic Dress  Category_Saree  \
0                0                 0                      0               0   
1                0                 0                      0               0   
2                0                 0                      0               0   
3                0                 0                      0               0   
4                0                 0                      0               0   

   Category_Set  Category_Top  Category_Western Dress  Category_kurta  \
0             1             0                       0               0   
1             0             0                       0               1   
2             0             0                       0               1   
3             0             0                       1               0   
4             0             1                       0               0   

   Sales Channel Amazon  Sales Channel Non-Amazon  promotion  
0                     1                         0          0  
1                     1                         0          1  
2                     1                         0          1  
3                     1                         0          0  
4                     1                         0          0

In [20]:
#now we fill null value on Amount column with 0
data_1.fillna(0,inplace=True)

In [21]:
#import package
import pyspark as ps
from pyspark.ml.linalg import Vector
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import BinaryClassificationEvaluator

#build session on Spark
spark = ps.sql.SparkSession.builder.master('local').appName('SuccessOrNo').getOrCreate()

#read pandas data frame to spark
sparkDF = spark.createDataFrame(data_1)

#print schema data
sparkDF.printSchema()

23/11/16 22:28:01 WARN Utils: Your hostname, Dwis-MacBook-Air.local resolves to a loopback address: 127.0.0.1; using 192.168.100.184 instead (on interface en0)
23/11/16 22:28:01 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/11/16 22:28:01 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
/Users/dwiputra/.pyenv/versions/3.11.6/lib/python3.11/site-packages/pyspark/sql/pandas/conversion.py:485: FutureWarning: is_datetime64tz_dtype is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.DatetimeTZDtype)` instead.
  if should_localize and is_datetime64tz_dtype(s.dtype) and s.dt.tz is not None:


root
 |-- ship-service-level: long (nullable = true)
 |-- Size: long (nullable = true)
 |-- Qty: long (nullable = true)
 |-- Amount: double (nullable = true)
 |-- ship-postal-code: long (nullable = true)
 |-- B2B: long (nullable = true)
 |-- fulfilled-by: long (nullable = true)
 |-- Success: long (nullable = true)
 |-- Fulfilment_Amazon: long (nullable = true)
 |-- Fulfilment_Merchant: long (nullable = true)
 |-- Category_Blouse: long (nullable = true)
 |-- Category_Bottom: long (nullable = true)
 |-- Category_Dupatta: long (nullable = true)
 |-- Category_Ethnic Dress: long (nullable = true)
 |-- Category_Saree: long (nullable = true)
 |-- Category_Set: long (nullable = true)
 |-- Category_Top: long (nullable = true)
 |-- Category_Western Dress: long (nullable = true)
 |-- Category_kurta: long (nullable = true)
 |-- Sales Channel Amazon: long (nullable = true)
 |-- Sales Channel Non-Amazon: long (nullable = true)
 |-- promotion: long (nullable = true)



Now we need to transform our data from all columns to 1 column of vector

In [22]:
#create a list of all column that will used for model feature
assemblerinput = ['ship-service-level', 'Size', 'Qty', 'Amount', 'ship-postal-code', 'B2B',
                  'fulfilled-by', 'Success', 'Fulfilment_Amazon', 'Fulfilment_Merchant',
                  'Sales Channel Amazon', 'Sales Channel Non-Amazon', 'Category_Blouse',
                  'Category_Bottom', 'Category_Dupatta', 'Category_Ethnic Dress', 'Category_Saree',
                  'Category_Set', 'Category_Top', 'Category_Western Dress', 'Category_kurta', 'promotion']

#we need to set maxToStringFields config because our vector quite long
spark.conf.set("spark.sql.debug.maxToStringFields", 100)

#run function to build vector and store it as feature column
vector = VectorAssembler(inputCols=assemblerinput,outputCol='features')
assembler = vector.transform(sparkDF)

#check the result
assembler.show(5)

23/11/16 22:28:10 WARN TaskSetManager: Stage 0 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.


+------------------+----+---+------+----------------+---+------------+-------+-----------------+-------------------+---------------+---------------+----------------+---------------------+--------------+------------+------------+----------------------+--------------+--------------------+------------------------+---------+--------------------+
|ship-service-level|Size|Qty|Amount|ship-postal-code|B2B|fulfilled-by|Success|Fulfilment_Amazon|Fulfilment_Merchant|Category_Blouse|Category_Bottom|Category_Dupatta|Category_Ethnic Dress|Category_Saree|Category_Set|Category_Top|Category_Western Dress|Category_kurta|Sales Channel Amazon|Sales Channel Non-Amazon|promotion|            features|
+------------------+----+---+------+----------------+---+------------+-------+-----------------+-------------------+---------------+---------------+----------------+---------------------+--------------+------------+------------+----------------------+--------------+--------------------+------------------------+

23/11/16 22:28:15 WARN PythonRunner: Detected deadlock while completing task 0.0 in stage 0 (TID 0): Attempting to kill Python Worker


In [23]:
#now we drop all column that we have used before
dataset_1 = assembler.drop('ship-service-level', 'Size', 'Qty', 'Amount', 'ship-postal-code', 'B2B',
                  'fulfilled-by', 'Fulfilment_Amazon', 'Fulfilment_Merchant',
                  'Sales Channel Amazon', 'Sales Channel Non-Amazon', 'Category_Blouse',
                  'Category_Bottom', 'Category_Dupatta', 'Category_Ethnic Dress', 'Category_Saree',
                  'Category_Set', 'Category_Top', 'Category_Western Dress', 'Category_kurta', 'promotion')

#check the result, now we only have 2 columns
dataset_1.show(5)

23/11/16 22:28:15 WARN TaskSetManager: Stage 1 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.


+-------+--------------------+
|Success|            features|
+-------+--------------------+
|      1|(22,[0,1,3,4,6,7,...|
|      1|(22,[0,1,2,3,4,6,...|
|      1|(22,[0,1,2,3,4,5,...|
|      1|(22,[0,1,3,4,6,7,...|
|      1|(22,[0,1,2,3,4,7,...|
+-------+--------------------+
only showing top 5 rows



23/11/16 22:28:19 WARN PythonRunner: Detected deadlock while completing task 0.0 in stage 1 (TID 1): Attempting to kill Python Worker


Now our data is ready to use as training and test a model of machine learning.

Because our model is binary classification, so we only need Logistic Classification (logistic regression in Spark).

In [24]:
#we split our data to 80% as training data and 20% as test data
(train,test) = dataset_1.randomSplit([0.8, 0.2], 48) #48 is seed that we use

#call Logistic Regression and fit training data
model_1 = LogisticRegression(featuresCol='features',labelCol='Success').fit(train)

23/11/16 22:28:19 WARN TaskSetManager: Stage 2 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.
23/11/16 22:28:21 WARN InstanceBuilder: Failed to load implementation from:dev.ludovic.netlib.blas.JNIBLAS
23/11/16 22:28:21 WARN TaskSetManager: Stage 3 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.
23/11/16 22:28:22 WARN TaskSetManager: Stage 4 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.
23/11/16 22:28:22 WARN TaskSetManager: Stage 5 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.
23/11/16 22:28:22 WARN TaskSetManager: Stage 6 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.
23/11/16 22:28:22 WARN TaskSetManager: Stage 7 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.
23/11/16 22:28:22 WARN TaskSetManager: Stage 8 

The training is done, let's check our result

In [25]:
result_1 = model_1.summary
print('Overall Accuracy Model 1 : ', (result_1.accuracy)*100, '%')

23/11/16 22:28:24 WARN TaskSetManager: Stage 28 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.


Overall Accuracy Model 1 :  100.0 %


Wow, our model is so good, it's mean our feature engineering is nice. Our model accuracy is 100%

Now we need to check with test data.

In [26]:
#do prediction with test data
predict = model_1.transform(test)

#call function to calculate accuracy
evaluator = BinaryClassificationEvaluator(labelCol='Success')

print('Accuracy Model 1 : ', (evaluator.evaluate(predict))*100, '%')

23/11/16 22:28:25 WARN TaskSetManager: Stage 30 contains a task of very large size (7188 KiB). The maximum recommended task size is 1000 KiB.


Accuracy Model 1 :  99.99979399911173 %


Our first model is so good and almost perfect prediction with 99% accuracy. But, we still need to check our model, does it overfitting or not with another dataset. But right now is Good enough to see our model perform so well.

## Machine Learning 2

Our second model is a model to forecasting sales transaction

Now we will prepare our data for creating machine learning forecasting.

First we need to slicing our data, we only need transaction data that is declared successful (has a value in the Amount column).

In [27]:
#to refresh our memory, lets check our dataset
sales_data_fix.head()

Order ID       Date                        Status  \
index                                                                 
0      405-8078784-5731545 2022-04-30                     Cancelled   
1      171-9198151-1101146 2022-04-30  Shipped - Delivered to Buyer   
2      404-0687676-7273146 2022-04-30                       Shipped   
3      403-9615377-8133951 2022-04-30                     Cancelled   
4      407-1069790-7240320 2022-04-30                       Shipped   

      Fulfilment Sales Channel  ship-service-level    Style              SKU  \
index                                                                          
0       Merchant      Amazon.in           Standard   SET389   SET389-KR-NP-S   
1       Merchant      Amazon.in           Standard  JNE3781  JNE3781-KR-XXXL   
2         Amazon      Amazon.in          Expedited  JNE3371    JNE3371-KR-XL   
3       Merchant      Amazon.in           Standard    J0341       J0341-DR-L   
4         Amazon      Amazon.in          Expedited  JNE3671  JNE3671-TU-XXXL   

            Category Size        ASIN Courier Status  Qty currency  Amount  \
index                                                                        
0                Set    S  B09KXVBD7Z            NaN    0      INR  647.62   
1              kurta  3XL  B09K3WFS32        Shipped    1      INR  406.00   
2              kurta   XL  B07WV4JV4D        Shipped    1      INR  329.00   
3      Western Dress    L  B099NRCT7B            NaN    0      INR  753.33   
4                Top  3XL  B098714BZP        Shipped    1      INR  574.00   

         ship-city   ship-state  ship-postal-code ship-country  \
index                                                            
0           MUMBAI  MAHARASHTRA          400081.0           IN   
1        BENGALURU    KARNATAKA          560085.0           IN   
2      NAVI MUMBAI  MAHARASHTRA          410210.0           IN   
3       PUDUCHERRY   PUDUCHERRY          605008.0           IN   
4          CHENNAI   TAMIL NADU          600073.0           IN   

                                           promotion-ids    B2B fulfilled-by  \
index                                                                          
0                                                    NaN  False    Easy Ship   
1      Amazon PLCC Free-Financing Universal Merchant ...  False    Easy Ship   
2           IN Core Free Shipping 2015/04/08 23-48-5-108   True          NaN   
3                                                    NaN  False    Easy Ship   
4                                                    NaN  False          NaN   

      Unnamed: 22 Success  
index                      
0             NaN     Yes  
1             NaN     Yes  
2             NaN     Yes  
3             NaN     Yes  
4             NaN     Yes

For forecasting we only need 2 column, "Date" and "Amount". So we will take just this 2 columns.

In [28]:
#select 2 column and store it on new variable and remove null value
data_forc = sales_data_fix[['Date','Amount']].copy().dropna()

#check result
display(data_forc.info())
data_forc.head()

<class 'pandas.core.frame.DataFrame'>
Index: 121149 entries, 0 to 128974
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   Date    121149 non-null  datetime64[ns]
 1   Amount  121149 non-null  float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 2.8 MB


None

Date  Amount
index                   
0     2022-04-30  647.62
1     2022-04-30  406.00
2     2022-04-30  329.00
3     2022-04-30  753.33
4     2022-04-30  574.00

Now we will not use spark to get it done because we need to plot our data, so we need to do on Pandas.

Next, because our data have is not daily amount of transaction so wee need to sum all amount on the day. We will group it by date and sum all amount value.

In [29]:
#group by date and sum amount column
data_forc = data_forc.groupby(['Date']).sum()
data_forc['Date'] = data_forc.index

#check result
display(data_forc.head())

Amount       Date
Date                             
2022-03-31   101683.85 2022-03-31
2022-04-01   865478.60 2022-04-01
2022-04-02   913101.53 2022-04-02
2022-04-03  1011763.38 2022-04-03
2022-04-04   882059.17 2022-04-04

Now we can plot out data

In [30]:
#import plotly
import plotly.express as px
import plotly.graph_objects as go

#plot chart
fig = px.line(data_forc,x='Date' ,y='Amount', title= 'Total Daily Transactions on Amazon.in')
fig.show()

Next, we will do faeture engineering. Because we cant use date as featue, so we need to collect from our date that can we use as feature.

we will create feature day, month, day of week, quarter, and does it holliday?\
we will use holidays package and use India holiday dates because we use dataset of India Amazon.

In [31]:
#import holidays package
import holidays

#set package as india
in_holiday = holidays.IN()

#collect information from date column
data_forc['day'] = data_forc['Date'].dt.day
data_forc['month'] = data_forc['Date'].dt.month
data_forc['dayofweek'] = data_forc['Date'].dt.day_of_week
data_forc['quarter'] = data_forc['Date'].dt.quarter
data_forc['holiday'] = data_forc['Date'].apply(lambda x : 1 if x in in_holiday else 0)

#check result
data_forc.head()

Amount       Date  day  month  dayofweek  quarter  holiday
Date                                                                      
2022-03-31   101683.85 2022-03-31   31      3          3        1        0
2022-04-01   865478.60 2022-04-01    1      4          4        2        0
2022-04-02   913101.53 2022-04-02    2      4          5        2        0
2022-04-03  1011763.38 2022-04-03    3      4          6        2        0
2022-04-04   882059.17 2022-04-04    4      4          0        2        0

Because we need to create model for forecasting, our data must be in series. So, we will split our train and test data using date.

In [32]:
#split data by date column, here we use 2022-06-19 as our treshold
train_for = data_forc.loc[data_forc['Date'] <= '2022-06-19'].reset_index(drop=True)
test_for = data_forc.loc[data_forc['Date'] >= '2022-06-19'].reset_index(drop=True)

#check the result
display(train_for.tail())
test_for.head()

Amount       Date  day  month  dayofweek  quarter  holiday
76  785089.32 2022-06-15   15      6          2        2        0
77  778813.50 2022-06-16   16      6          3        2        0
78  747349.10 2022-06-17   17      6          4        2        0
79  741010.38 2022-06-18   18      6          5        2        0
80  729840.60 2022-06-19   19      6          6        2        0

Amount       Date  day  month  dayofweek  quarter  holiday
0  729840.60 2022-06-19   19      6          6        2        0
1  756764.91 2022-06-20   20      6          0        2        0
2  804470.44 2022-06-21   21      6          1        2        0
3  778915.31 2022-06-22   22      6          2        2        0
4  654729.48 2022-06-23   23      6          3        2        0

Ok, now we have train and test data and it looks good. Now, we will train our model, here we use eXtreme Gradient Boosting (XGB) regressor with 50 estimator.

In [33]:
#import eXtreme Gradient Boosting package
import xgboost as xgb

#split data with X as feature and y as target
X_train = train_for.drop(['Amount','Date'],axis=1)
y_train = train_for[['Amount','Date']]
X_test = test_for.drop(['Amount','Date'],axis=1)
y_test = test_for[['Amount','Date']]

#create model_2 and fit training model
model_2 = xgb.XGBRegressor(n_estimators = 50)
model_2.fit(X_train,y_train['Amount'], eval_set=[(X_train,y_train['Amount']), (X_test,y_test['Amount'])], verbose=True)

[0]	validation_0-rmse:119921.05759	validation_1-rmse:197070.30779
[1]	validation_0-rmse:99712.51628	validation_1-rmse:181766.85662
[2]	validation_0-rmse:84485.54323	validation_1-rmse:176271.46704
[3]	validation_0-rmse:71079.75132	validation_1-rmse:162154.27785
[4]	validation_0-rmse:61317.66179	validation_1-rmse:158816.73814
[5]	validation_0-rmse:52883.79465	validation_1-rmse:151079.57269
[6]	validation_0-rmse:45705.83964	validation_1-rmse:149548.73172
[7]	validation_0-rmse:40375.29201	validation_1-rmse:149145.40957
[8]	validation_0-rmse:36264.87026	validation_1-rmse:148721.52719
[9]	validation_0-rmse:32673.02649	validation_1-rmse:149963.82707
[10]	validation_0-rmse:30187.50536	validation_1-rmse:150115.79471
[11]	validation_0-rmse:27980.64004	validation_1-rmse:150375.43783
[12]	validation_0-rmse:25321.76853	validation_1-rmse:151395.42372
[13]	validation_0-rmse:22543.90899	validation_1-rmse:151947.57853
[14]	validation_0-rmse:20862.44980	validation_1-rmse:151556.20734
[15]	validation_0-r

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=50, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

here we got our model get RMSE = 149662, this score tells us that the average deviation between the predicted amount of daily transaction and the actual amount of daily transaction is 149662 INR.

from these results it can be said that our 2nd model has not been able to predict the Amount of transaction very well.

This is very likely to happen because the amount of transaction data in the training dataset is only transaction data for 2.5 months. therefore the model we make is less trained so that the ability is inadequate.

Next, let's look at the data from the prediction results of the 2nd model against the testing data.

In [34]:
#do prediction to X test data and store the result to y feature data. So that we can easily compare the results of the two models.
y_test['Amount - Predict'] = model_2.predict(X_test)

#calculate deviation between predicted and actual Amount
y_test['deviation'] = abs(y_test['Amount - Predict']-y_test['Amount'])

#check result
y_test.head()

/var/folders/0j/0hdq0xy56l92jvr6tsb_35rc0000gn/T/ipykernel_29565/1143487589.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/var/folders/0j/0hdq0xy56l92jvr6tsb_35rc0000gn/T/ipykernel_29565/1143487589.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Amount       Date  Amount - Predict    deviation
0  729840.60 2022-06-19       731417.4375    1576.8375
1  756764.91 2022-06-20       810309.6250   53544.7150
2  804470.44 2022-06-21       759194.9375   45275.5025
3  778915.31 2022-06-22       787935.7500    9020.4400
4  654729.48 2022-06-23       756131.7500  101402.2700

In [35]:
#create and show figure to multiple line plot
fig = go.Figure()
fig.add_trace(go.Scatter(x=y_test['Date'], y=y_test['Amount'],
                    mode='lines',
                    name='Actual',))
fig.add_trace(go.Scatter(x=y_test['Date'], y=y_test['Amount - Predict'],
                    mode='lines',
                    name='Prediction'))
fig.update_layout(title='Plot of Actual Amount and Prediction Amount')
fig.show()

#create figure of deviation of actual amount and prediction amount
fig = px.line(y_test,x='Date' ,y='deviation', title= 'Deviation of Actual Amount and Prediction Amount')
fig.show()

from the above we know our model not always give good prediction and we need more data to fix it. But now, lets do some forecasting.

Our data is end at 29 Juny 2022 and that day the amount of date is like anomaly data, so we will forcast to 2 month later from 29 Juny 2022.

In [36]:
#create dateframe with range time
forcast = pd.date_range(start="2022-06-29",end="2022-08-31").to_frame(name='Date').reset_index(drop=True)

#do feature engineering same as before
forcast['day'] = forcast['Date'].dt.day
forcast['month'] = forcast['Date'].dt.month
forcast['dayofweek'] = forcast['Date'].dt.day_of_week
forcast['quarter'] = forcast['Date'].dt.quarter
forcast['holiday'] = forcast['Date'].apply(lambda x : 1 if x in in_holiday else 0)

#check result
forcast.head()

Date  day  month  dayofweek  quarter  holiday
0 2022-06-29   29      6          2        2        0
1 2022-06-30   30      6          3        2        0
2 2022-07-01    1      7          4        3        0
3 2022-07-02    2      7          5        3        0
4 2022-07-03    3      7          6        3        0

Now we can do forecasting to the 2 next month

In [37]:
#Predict and store in new column
forcast['Amount - Predict'] = model_2.predict(forcast.drop(['Date'],axis=1))

#check result
forcast.head()

Date  day  month  dayofweek  quarter  holiday  Amount - Predict
0 2022-06-29   29      6          2        2        0       827855.6875
1 2022-06-30   30      6          3        2        0       803789.5625
2 2022-07-01    1      7          4        3        0       782451.2500
3 2022-07-02    2      7          5        3        0       807819.2500
4 2022-07-03    3      7          6        3        0       815056.0625

Because we start forecasting from 2022-06-29, we need to remove 2022-06-29 on actual dataset

In [38]:
#slice by indexdate
new_slice = data_forc.loc[data_forc.index != '2022-06-29']

#check result
new_slice.tail()

Amount       Date  day  month  dayofweek  quarter  holiday
Date                                                                     
2022-06-24  630349.29 2022-06-24   24      6          4        2        0
2022-06-25  654234.58 2022-06-25   25      6          5        2        0
2022-06-26  773610.02 2022-06-26   26      6          6        2        0
2022-06-27  714124.67 2022-06-27   27      6          0        2        0
2022-06-28  772085.53 2022-06-28   28      6          1        2        0

In [39]:
#Create figure to see the prediction result
fig = go.Figure()
fig.add_trace(go.Scatter(x=new_slice['Date'], y=data_forc['Amount'],
                    mode='lines',
                    name='Actual',))
fig.add_trace(go.Scatter(x=forcast['Date'], y=forcast['Amount - Predict'],
                    mode='lines',
                    name='Prediction'))
fig.update_layout(title='forecasting Amount of Daily Transactions on Amazon.in')
fig.update_traces(connectgaps=True)
fig.show()

Here our result in graph.

It still looks good. We still see the spikes that may occur on certain dates and nominally this is still in the right area (not much increase or decrease) as the data is only 2 months apart. But, it might not be correct in 90% as we have a pretty big RMSE.

Here we see missing line in 28 to 29 June 2022. it is because we are graphing two separate variables, the system does not recognize them as one line. However, this is not a problem because we will display this data in the tableau.

So, we need to save our prediction result to csv file.

In [40]:
#first add flag to each dataframe
new_slice['flag'] = 'Actual'
forcast['flag'] = 'Prediction'
display(new_slice.head())
forcast.head()

/var/folders/0j/0hdq0xy56l92jvr6tsb_35rc0000gn/T/ipykernel_29565/3616305394.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Amount       Date  day  month  dayofweek  quarter  holiday  \
Date                                                                         
2022-03-31   101683.85 2022-03-31   31      3          3        1        0   
2022-04-01   865478.60 2022-04-01    1      4          4        2        0   
2022-04-02   913101.53 2022-04-02    2      4          5        2        0   
2022-04-03  1011763.38 2022-04-03    3      4          6        2        0   
2022-04-04   882059.17 2022-04-04    4      4          0        2        0   

              flag  
Date                
2022-03-31  Actual  
2022-04-01  Actual  
2022-04-02  Actual  
2022-04-03  Actual  
2022-04-04  Actual

Date  day  month  dayofweek  quarter  holiday  Amount - Predict  \
0 2022-06-29   29      6          2        2        0       827855.6875   
1 2022-06-30   30      6          3        2        0       803789.5625   
2 2022-07-01    1      7          4        3        0       782451.2500   
3 2022-07-02    2      7          5        3        0       807819.2500   
4 2022-07-03    3      7          6        3        0       815056.0625   

         flag  
0  Prediction  
1  Prediction  
2  Prediction  
3  Prediction  
4  Prediction

In [41]:
#set all columns in same order and same column name
new_slice = new_slice[['Date','day','month','dayofweek','quarter','holiday','Amount','flag']]
forcast.columns = ['Date','day','month','dayofweek','quarter','holiday','Amount','flag']

#check result
display(new_slice.tail())
forcast.head()

Date  day  month  dayofweek  quarter  holiday     Amount  \
Date                                                                        
2022-06-24 2022-06-24   24      6          4        2        0  630349.29   
2022-06-25 2022-06-25   25      6          5        2        0  654234.58   
2022-06-26 2022-06-26   26      6          6        2        0  773610.02   
2022-06-27 2022-06-27   27      6          0        2        0  714124.67   
2022-06-28 2022-06-28   28      6          1        2        0  772085.53   

              flag  
Date                
2022-06-24  Actual  
2022-06-25  Actual  
2022-06-26  Actual  
2022-06-27  Actual  
2022-06-28  Actual

Date  day  month  dayofweek  quarter  holiday       Amount        flag
0 2022-06-29   29      6          2        2        0  827855.6875  Prediction
1 2022-06-30   30      6          3        2        0  803789.5625  Prediction
2 2022-07-01    1      7          4        3        0  782451.2500  Prediction
3 2022-07-02    2      7          5        3        0  807819.2500  Prediction
4 2022-07-03    3      7          6        3        0  815056.0625  Prediction

In [42]:
#append 2 dataframe
result = pd.concat([new_slice, forcast], ignore_index=True)
display(result.head())

#save as csv
result.to_csv('forcast_result.csv')

Date  day  month  dayofweek  quarter  holiday      Amount    flag
0 2022-03-31   31      3          3        1        0   101683.85  Actual
1 2022-04-01    1      4          4        2        0   865478.60  Actual
2 2022-04-02    2      4          5        2        0   913101.53  Actual
3 2022-04-03    3      4          6        2        0  1011763.38  Actual
4 2022-04-04    4      4          0        2        0   882059.17  Actual

## Conclusion

In this project, we use sales transaction data from the world's leading e-commerce company Amazon, which we focus on Amazon India on transactions of various types of clothing.

In this data of clothing sales transactions in India we found a problem in the success rate of each transaction made. We never seem to know when a transaction will end successfully or fail. Therefore, in this project we attempt to create a model that can provide predictions of the success or failure of transactions that occur. This model was created with the hope of providing input to the transaction security system to provide preventive steps before the transaction is actually carried out.

In this first model, we managed to get very satisfying results at 99% accuracy (based on testing data). Perhaps this accuracy is too high and should be tested on more diverse test data and with a larger number. However, this very high accuracy can also be caused by failed transaction patterns that are very easy to recognize by machine learning. However, we can be proud that only by using the logistic classifier model, we managed to get a very good machine learning model performance.

Next, we also built a forecasting model to predict the amount of hasian transactions that occur on Amazon India. This model was also successful but with an RMSE value of 149,662 INR per total daily transaction amount. This can happen because the data we have is only transaction data for 2.5 months from May to July 2022. Even so, the forecasting results produced by this 2nd model look quite good and promising when we do forecasting until the next 2 months, August 2022. This forcesting model does not use the time series model because it considers the amount of data is quite small, while the time series requires the amount of data up to an annual scale. Therefore, in this model we use eXtreme Gradient Boosting (XGB) regressor with 50 estimators.


Thank you.